### ISSUE TITLE
Execute badgerauth test plan

### ISSUE THREAD

**Issue by ferristocrat** [storj/gateway-mt#219]
**Goal**
Execute test plan (below) created as part of https://github.com/storj/gateway-mt/issues/193

**Acceptance Criteria**

1. Each test case is executed
2. The status of each test case is added as passing or failing
3. For each failing test case, a bug is created in the gateway-mt repository

**Test Cases**

Test Scenario | Test Case | Description | Comments
-- | -- | -- | --
Security | Mutual Authentication | It should not be possible to start cluster without mutual authentication, have a test for trying to start a cluster without mutual authentication |  
  | Security Certificates- Generation | When generating the security certificates there should be a CA certificate, Server certificate and Key for server certificate |  
  | Security Certificates- Management | Certificates should be managed by a system and not a user (improving security) |  
  | Security Certificates- Rotation | Certificates should be rotated frequently for operational security; whether this is done manual or automated |  
  | Security Certificates- Expiration | Certificates should have an expiration, default 365 days, this is because one of the core features of SSLs is server authentication and as expirations carry validity periods they are used to establish secure connections |  
  | Directory (certs) | There should be a directory that contains CA and node certificates, also from this directory relevant files should be upload-able to nodes |  
  | Directory (safe-directory) | There should be a safe directory where CA keys can be stored safely (safe directory) and backed up, otherwise if its not stored safely and then gets compromised damage done can be substantial, and if lost it will be impossible to add new nodes to a cluster The CA key should also never be uploaded to nodes, otherwise, it defeats the purpose of keeping it safe and secret | First layer of security for CA key
  | CA key (Multiperson Control) | There should be multiple objects required to activate and use CA private key, this will add another layer of security ((for example)as these multiple objects themselves can be separated and securely stored in a hardware security module or through locks and authentications) | Second layer of security for CA key
  | Network Security | Because it was decided to not make it impossible for nodes to ask for data it does not know about, there should be some kind of network policy or security to prevent rogue nodes from communicating with other nodes in a cluster or there could be a compromised cluster just because of one node that is affected |  
  | Cluster Differences | Misalignments in cluster nodes software shouldn't hinder security and or performance |  
  | Encryption | Data that is not contained in a safe directory should be encrypted for security purposes |  
  | Backups | Backups should be just local replicas compressed into Storj DCS after specified intervals | Moved from Design
Monitoring | Metrics | The new auth database should report metrics/events prefixed with 'as_badgerauth_' |  
  | Logs/Auditing 1 | There should be obtainable logs containing troubleshooting/helpful information at DEBUG level; f.e if there was a problem with a cluster, where did it occur, how/when did it happen, what was observed etc. | this can be implemented in the future since at the moment basic Info and metrics should be enough
  | Logs/Auditing 2 | If the above is implemented, then as a further improvement, there should be some type of Audit Policy where logs can be grouped in terms of severity(ignore or fix), priority, metadata info, requests etc. |  
  | Admin Tool | There should be an admin tool to query/monitor the badgerauth database (CLI tool) |  
  | Alerts | If there are any important logs that hint a problem/issue there should be an alert attached to it |

**Comment by storj-antonio** [comment]
There are some other tests that should be included in this plan that @nadimhq wrote up in a separate ticket. I'll get the full list over and prune out the ones that @amwolff indicated were out of scope for the authservice.

**Comment by storj-antonio** [comment]
Test Scenario | Test Case | Description | Comments
-- | -- | -- | --
Latency | Read or Write Latency | If we have read or write latency it shouldn't be affected by cross-regional latency | **from design blueprint- latency range should be a couple of milliseconds
  | Replication Latency | If we have replication latency, then it should be configurable | **latency range should be a couple of seconds
ACID+ | Database Consistency | If we have an auth database, it should be eventually consistent and shouldn't read stale data according to defined rules |  
  | Database Availability | Even if other authservices/regions are down, if we have any online authservice it should be able to respond to data from the DB |  
  | Database Isolation | If we have an auth database, then having multiple transactions in a database should not hinder it and should work as if they were executed one after the other | *from the design blueprint, Load- auth database doesn't do more than 1 request per second on average for both reads and writes however it can spike to 25 per second and is able to handle that
  | Database Durability 1 | If we have an auth database, then it should be durable and no external factors should hinder it | *from the design blueprint, we can ensure to never lose a record before it gets replicated to other nodes by having it write to two nodes synchronously as an example
  | Database Durability 2 | If we have two nodes synchronously written with the same record, then this should increase durability of writes to ensure we never lose a record before it gets replicated to other nodes | -if previously, mentioned durability change is used
  | Database Durability 3 | If a record is inserted into one authservice, then the authservice responds with mapped credentials, crashes, and then comes back up later one should be able to retrieve the inserted record using said mapped credentials |  
  | Database Availability/Durability | Even if other authservices/regions are down, if we have any online authservice it should be able to respond to data from the DB |  
  | Database Atomicity | Records should be serialized, contain information such as (record data, denormalized information from access grant, sensitive data, invalid tracking, State and synchronization-related data) and saved in Protocol buffers binary format (any missing data should invalidate?) |  
  | Data Integrity 1 | Records should be able to be added, retrieved, updated or deleted | **explained in KV interface implementation
  | Data Integrity 2 | A record should not be retrievable immediately after insertion into an authservice until only after replication |  
Design | Generate Authservice ID | If we generate an ID for authservice at the start of an instance, then we should be able to avoid the case of human error (when someone puts up a cluster with the two same node IDs) |  
  | Alpha Implementation | At the start we should implement Put, Get, DeleteUnused and the replication mechanism |  
  | External Tool | For Deletion/invalidation/modification we should let that be handled by the auth-admin CLI tool that connects to all nodes if needed |  
  | Unique ID 1 | Each node should have its own globally unique ID, test what happens with a node with a different ID than in the database |  
  | Unique ID 2 | Each node should have its own globally unique ID, test what happens if two nodes are trying to be started with the same ID |  
  | Replication Loop | Every node should run a replication loop, which should call replicate on all nodes that node knows in a fixed interval |  
  | Operations | Operations can be applied in any order to form the retrieval part of a local replica and should always result in a consistent view of the database |  
  | Adding New Nodes to Cluster | Adding a new node to the cluster should just update the list of nodes it knows about and adding it to the list of nodes other nodes know about |  
  | Deleting Nodes from Cluster | Eventually, nodes will retire and these nodes should be deleted from the list of nodes other nodes know about so there won't be complications from being unable to connect to said retired node |  
  | Migration | Migration could be triggered by a config flag, and as a background job it should move all data from CockroachDB to the new database, depending on the request (PUT- save record to new database) (GET- checks if new db contains the record and if not falls back to CockroachDB) |  
Implementation | Badger DB | BadgerDB should be used for the underlying key-value DB on every node |  
  | DRPC | DRPC should be used for replication-related communication |  
  | Record | Records should be serialized, contain information such as (record data, denormalized information from access grant, sensitive data, invalid tracking, State and synchronization-related data) and saved in Protocol buffers binary format |  
  | Replication Log Entry | With the replication log entry it can be key-only and may take this form- authservice_id/counter/encryption_key_hash/operation or it can also be a protobuf saved under authservice_id/counter |  
Key-Value Interface Implementation | Get | Get should retrieve record from KV store and then checks if it is valid and not in a temporary DELETED state, it returns nil if it doesnt exist and if its invalid it should explain in an error why |  
  | Put | Put should save record under encryption_key_hash and in addition create a corresponding replication log entry with the CREATED operation, if the key already exists then there should be an error for this case |  
  | Ping | Ping attempts a DB roundtrip, but in the case of BadgerDB which would be embedded into the authservice it shouldn't be able to ping and thus this method should always return a nil error | *from design blueprint, can implement adding a test record and deleting it for Ping operation to ensure things work correctly
  | Close | Close should close the database and thus close the replication chore |  
Replication Chore Implementation | Replication Loop | Every node should run a replication loop, which means that all nodes should call Replicate on every node it knows in a fixed interval |  
  | Replicate Node | The node that calls replicate should gather the highest available counters for all nodes it knows about from its replication log and send them in the request |  
  | Responding Node | Nodes that recieve a request from a node that calls replicate, should gather records and replication log entries above the counters from the request and send them in a batch that is configurable up to 10,000 records |  
  | Node Recieving the Batch | A node that recieves said batch should add the recieved replication log entries and records to its local replica. Also records should be mutated with respect to the Design/Replication Algorithm f.e uf tge record entry has a TTL that should be added as well. Lastly, the created_at field should be updated asl well with the current time of the node recieving the response |  
  | Missed Log Entries | No log entries should be missed, since when a syncing node asks for entries it will look at the corresponding available records and if they don't match then the node will be out of sync |  
  
  Test Scenario | Test Case | Description | Comments
-- | -- | -- | --
Security | Mutual Authentication | It should not be possible to start cluster without mutual authentication, have a test for trying to start a cluster without mutual authentication |  
  | Security Certificates- Generation | When generating the security certificates there should be a CA certificate, Server certificate and Key for server certificate |  
  | Security Certificates- Management | Certificates should be managed by a system and not a user (improving security) |  
  | Security Certificates- Rotation | Certificates should be rotated frequently for operational security; whether this is done manual or automated |  
  | Security Certificates- Expiration | Certificates should have an expiration, default 365 days, this is because one of the core features of SSLs is server authentication and as expirations carry validity periods they are used to establish secure connections |  
  | Directory (certs) | There should be a directory that contains CA and node certificates, also from this directory relevant files should be upload-able to nodes |  
  | Directory (safe-directory) | There should be a safe directory where CA keys can be stored safely (safe directory) and backed up, otherwise if its not stored safely and then gets compromised damage done can be substantial, and if lost it will be impossible to add new nodes to a cluster The CA key should also never be uploaded to nodes, otherwise, it defeats the purpose of keeping it safe and secret | First layer of security for CA key
  | CA key (Multiperson Control) | There should be multiple objects required to activate and use CA private key, this will add another layer of security ((for example)as these multiple objects themselves can be separated and securely stored in a hardware security module or through locks and authentications) | Second layer of security for CA key
  | Network Security | Because it was decided to not make it impossible for nodes to ask for data it does not know about, there should be some kind of network policy or security to prevent rogue nodes from communicating with other nodes in a cluster or there could be a compromised cluster just because of one node that is affected |  
  | Cluster Differences | Misalignments in cluster nodes software shouldn't hinder security and or performance |  
  | Encryption | Data that is not contained in a safe directory should be encrypted for security purposes |  
  | Backups | Backups should be just local replicas compressed into Storj DCS after specified intervals | Moved from Design
Monitoring | Metrics | The new auth database should report metrics/events prefixed with 'as_badgerauth_' |  
  | Logs/Auditing 1 | There should be obtainable logs containing troubleshooting/helpful information at DEBUG level; f.e if there was a problem with a cluster, where did it occur, how/when did it happen, what was observed etc. | this can be implemented in the future since at the moment basic Info and metrics should be enough
  | Logs/Auditing 2 | If the above is implemented, then as a further improvement, there should be some type of Audit Policy where logs can be grouped in terms of severity(ignore or fix), priority, metadata info, requests etc. |  
  | Admin Tool | There should be an admin tool to query/monitor the badgerauth database (CLI tool) |  
  | Alerts | If there are any important logs that hint a problem/issue there should be an alert attached to it |
